src/Ucommerce.Transactions.Payments.Freepay/FreepayPaymentMethodService.cs

```diff
@@ -0,0 +1,397 @@
+
+            responseCode = response.StatusCode;
+
+            return content;
+        }
+
+        private string GeneratePaymentLink(PaymentRequest paymentRequest)
+        {
+            string apiKey = paymentRequest.PaymentMethod.DynamicProperty<string>().ApiKey.ToString();
+            string acceptUrlConfig = paymentRequest.PaymentMethod.DynamicProperty<string>().AcceptUrl;
+            string declineUrlConfig = paymentRequest.PaymentMethod.DynamicProperty<string>().CancelUrl;
+            bool testMode = paymentRequest.PaymentMethod.DynamicProperty<bool>().TestMode;
+
+            var twoLetterLanguageCode = GetFourLetterLanguageCode(paymentRequest);
+            var amount = paymentRequest.Payment.Amount.ToCents().ToString();
+            var acceptUrl = new Uri(_absoluteUrlService.GetAbsoluteUrl(acceptUrlConfig)).AddOrderGuidParameter(paymentRequest.Payment.PurchaseOrder).ToString();
+            var declineUrl = new Uri(_absoluteUrlService.GetAbsoluteUrl(declineUrlConfig)).AddOrderGuidParameter(paymentRequest.Payment.PurchaseOrder).ToString();
+            var callbackUrl = _callbackUrl.GetCallbackUrl("(auto)", paymentRequest.Payment);
+
+            HttpClient client = new HttpClient();
+            client.DefaultRequestHeaders.Add("Authorization", apiKey);
+
+            var values = new Dictionary<string, object>
+            {
+                { "Amount", amount },
+                { "Currency", paymentRequest.Amount.CurrencyIsoCode },
+                { "OrderNumber", paymentRequest.PurchaseOrder.Id },
+                { "CustomerAcceptUrl", acceptUrl },
+                { "CustomerDeclineUrl", declineUrl },
+                { "ServerCallbackUrl", callbackUrl },
+                { "BillingAddress", GetAddressJson(paymentRequest.PurchaseOrder, true) },
+                { "ShippingAddress", GetAddressJson(paymentRequest.PurchaseOrder) },
+                { "SaveCard", false },
+                { "EnforceLanguage", twoLetterLanguageCode },
+                { "Client", new Dictionary<string, object> {
+                    { "CMS", new Dictionary<string, object> { { "Name", "Umraco" }, { "Version", "8" } } },
+                    { "Shop", new Dictionary<string, object> { { "Name", "uCommerce" }, { "Version", "9.4.2" } } },
+                    { "Plugin", new Dictionary<string, object> { { "Name", "Freepay" }, { "Version", "1.0" } } },
+                    { "API", new Dictionary<string, object> { { "Name", "Freepay" }, { "Version", "1.0" } } },
+                } },
+            };
+
+            if (testMode)
+            {
+                values.Add("Options", new Dictionary<string, object> { { "TestMode", true } });
+            }
+
+            var response = client.PostAsync("https://gw.freepay.dk/api/payment", new StringContent(JsonConvert.SerializeObject(values), Encoding.UTF8, "application/json")).Result;
+            string responseString = response.Content.ReadAsStringAsync().Result;
```

Comment:
.Result is not recommended, because of Threadpool Starvation. So if you can use another client maybe consider that.

if it throws an exception you will get an aggregatedexception, if you do `.GetAwaiter().GetResult()`instead it will be easier to debug.

Reply:
Change applied in revision 05018b322f5bf7ec2f35e22e57c60db974604361